### ISSUE TITLE
[GH Issue Summarization] Hyperparameter tuning(grid search)?

### ISSUE THREAD

**Issue by jlewi** [kubeflow/examples#37]
@hamelsmu @ankushagarwal 

For the GitHub issue summarization model would it make sense to do a simple grid search for hyperparameter tuning?

I think this would be pretty straightforward to implement

- Create a simple Python program(controller) to do a grid search
   - Launch N TFJobs at a time and wait to complete.
- Controller can store information in a file on a PD for resilience
   - SQLLite might be a good choice

- Run the controller as a K8s Job.

I think my main question, does the model have hyperparameters worth tuning? Are through suitable metrics for deciding which model is best?

**Comment by hamelsmu** [comment]
I don' think you have to go crazy with hyper-parameter tuning in this example.  The main hyperparameters worth tuning are:

- Learning Rate
- Dimensionality of the encoder, decoder and embedding which i set to the same value for simplicity -- in my tutorial code this variable is called `latent_dim` . 

If you really want to show off hyper-parm tuning, maybe only tune `latent_dim`.  

As far as model evaluation, we could use BLEU Score https://www.nltk.org/_modules/nltk/translate/bleu_score.html.   However I haven't built that into my tutorial.  For simplicity you could evaluate the model according to the loss metric `categorical_cross_entropy` but in this case you want to find the lowest dimensionality that maintains a good loss. 

In my experience, I kind of still play around with tuning of neural networks manually unlike your conventional ML methods (random forest, etc. ) just because there are so many knobs to tune.  You could also skip Hyper - Param tuning entirely if you want, depending upon the effort.  

I hope this information is helpful.  Please ping me with any additional questions.

**Comment by jlewi** [comment]
Thats great thank you so much.

**Comment by aronchick** [comment]
This is really cool Hamel! Any chance you'd like to work on this and/or
have the time? :)

**Comment by hamelsmu** [comment]
@aronchick Absolutely, would be more than happy to add this.  Let me know how you prefer I add this into the example pipeline ya'll are building and I'll submit a PR

**Comment by hamelsmu** [comment]
Hello folks, I spent some time thinking about this - I don't know if you want to do hyper-parameter tuning on this example, because it would take a really long time and might detract from the experience of learning Kubeflow.  I would suggest considering skipping the hyper param tuning, but adding in model evaluation on a holdout set.  

I added the model evaluation on the holdout set in the Jupyter Notebook.  Made two PR's (1) to update the notebook and (2) to fix a mistake I found for the code that evaluates the model.   

Let me know your thoughts, happy to help in any other way.

**Comment by jlewi** [comment]
Why do you think it would distract from learning Kubeflow?

Why can't we structure the example/tutorial as a sequence of increasingly complex modules? So the fist most basic module wouldn't include hyperparameter tuning but another module might.

**Comment by hamelsmu** [comment]
We can totally do hyper parameter tuning if you think it’s feasible.  I was just thinking that hyperparam tuning can take a really long time, that’s all.  

On a Tesla P100 Card, sufficient training takes about 10 minutes.  Assuming a grid of 5x5 parameters (learning rate, Latent_dim) this would take > 1 hour.  Do you think that would be a deterrent for some people?  I wonder what kind of hardware people will use for the tutorial?

Again I’m open to suggestions - curious to know what you think about it.

**Comment by jlewi** [comment]
@hamelsmu An hour isn't a long time at all; especially for a grid search since we can parallelize it.

I think its  very useful to have examples that run at scale. It will be a deterrent for some people but having examples that we can run and validate at scale is very useful.

**Comment by texasmichelle** [comment]
Do we still want to add this? A Katib extension could make sense. If not, let's close.

**Comment by hamelsmu** [comment]
fwiw the language model required tuning at massive scale https://github.com/machine-learning-apps/IssuesLanguageModel/tree/master/hyperparam_sweep  can use this as a template if you like!
